Update modern platform to 1.21.11

calcastor opened the pull request:
~~Not certain about what I have done for game rules here.~~

calcastor committed:
Update modern platform to 1.21.11

Signed-off-by: BT (calcastor/mame) <43831917+calcastor@users.noreply.github.com>

TTtie committed:
refactor: redo the gamerule system

Refactors the game rule system to support both 1.21.11 and 1.8.9 game
rules in both environments. Supplying both the legacy and the modern
game rule will result in an error. It also issues warnings to the
operators when trying to use a game rule that the server doesn't
suppport.

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
fix: port the block desync fix over

The issue is way worse under 1.21.11, now causing players to not be able to interact with the world at all.

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
feat: add spears into the weapon list

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on platform/platform-modern/src/main/java/tc/oc/pgm/platform/modern/listeners/ModernListener.java:
```diff
@@ -107,14 +110,21 @@ public void onPlayerTeleport(final PlayerTeleportEvent event) {
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    });
```
There's a `markClientUnloadedAfterDeath` and a `restartClientLoadTimerAfterRespawn`, maybe just use that to mark the player as respawned right after dead instead of reflection

TTtie replied:
None of the methods specified do the thing I want it to do.

- `markClientUnloadedAfterDeath` is the very method blocking interaction (see `ServerPlayer#die(DamageSource)`)
- `restartClientLoadTimerAfterRespawn` unmarks `ServerGamePacketListenerImpl#waitingForRespawn` but also adds the 60-tick client world load timeout (that will never happen as respawn screens are blocked), reintroducing #1593.

Pablete1234 commented on platform/platform-modern/src/main/java/tc/oc/pgm/platform/modern/modules/gamerules/ModernRemappedGameRule.java:
```diff
@@ -0,0 +1,45 @@
+        return Result.ok(handle.getType().cast(parsedValue));
+      } catch (ClassCastException e) {
+        return Result.err(e);
+      }
```
```suggestion
      return Result.ok(handle.getType().cast(parsedValue));
```
given it's already wrapped in a catch for throwable that has the exact same handling, seems these would be identical behaviors.

Pablete1234 commented on platform/platform-sportpaper/src/main/java/tc/oc/pgm/platform/sportpaper/modules/gamerules/SpGameRules.java:
```diff
@@ -0,0 +1,106 @@
+@Supports(SPORTPAPER)
+@NullMarked
+public class SpGameRules implements tc.oc.pgm.util.bukkit.GameRules.GameRuleRegistry {
+  protected static final @Unmodifiable Map<String, String> DEFAULTS;
```
Seems like DEFAULTS should just be a field populated in the GameRule (ie SpKVGameRule or whatever) object with whatever its default is

Pablete1234 commented on util/src/main/java/tc/oc/pgm/util/bukkit/GameRule.java:
```diff
@@ -0,0 +1,59 @@
+   * @param world The world to set the game rule in
+   * @param value The string representation of the value to set, or null to reset to default
+   */
+  void setFromString(World world, @Nullable String value);
```
I think this method should go, strings for gamerules are a thing of the past, our api should just pretend they're always `T` and legacy can do whatever it needs to to turn them back into strings internally if needed

Pablete1234 commented on util/src/main/java/tc/oc/pgm/util/bukkit/GameRule.java:
```diff
@@ -0,0 +1,59 @@
+   * @param value The string value to parse
+   * @return A Result containing the parsed value or an error if parsing failed
+   */
+  Result<T, ?> tryParse(String value);
```
I question the use of `Result<T, ?>`, it's very functional-based but in java the more traditional approach would simply be:

`T tryParse(String value) throws X` and make the caller try-catch for it, instead of expecting them to `switch` over the result.

I'm fine with either, just just leaving it as food for thought.

Pablete1234 commented on core/src/main/java/tc/oc/pgm/gamerules/GameRulesModule.java:
```diff
@@ -30,33 +35,59 @@ public GameRulesMatchModule createMatchModule(Match match) {
-          }
+            throw new InvalidXMLException(
+                "Missing value for game rule " + ruleName, gameRuleElement);
+          } else {
```
consider using an early return to avoid the whole method getting yet another indent from the `else` block:
```java
if (value == null)
  throw new InvalidXMLException("Missing value for game rule " + ruleName, gameRuleElement);

GameRule<?> rule = GameRules.getByName(ruleName);
// ...
```

calcastor committed:
Address a couple review comments

Signed-off-by: BT (calcastor/mame) <43831917+calcastor@users.noreply.github.com>

TTtie committed:
refactor: address review comments

- removed Result class and ported code over to the more idiomatic throwing style
- removed GameRule#setFromString
- moved default game rule value into SpKVGameRule

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
refactor: don't use NMS in ModernMiscUtil#spawnPotion

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
refactor: use Paper's Adventure integration instead of MinecraftComponentSerializer

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
refactor: don't use NMS in ModernNMSHacks#skipFireworksLaunch

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/gamerules/GameRulesModule.java:
```diff
@@ -78,12 +77,12 @@ public GameRulesModule parse(MapFactory factory, Logger logger, Document doc)
+          } catch (Throwable e) {
+            throw new InvalidXMLException(
+                "Failed to parse game rule value for " + rule.name() + ": " + e.getMessage(),
+                gameRuleElement);
```
You're missing passing `e` as the cause for the invalid xml exception
```suggestion
            throw new InvalidXMLException(
                "Failed to parse game rule value for " + rule.name() + ": " + e.getMessage(),
                gameRuleElement,
                e);
```

TTtie committed:
chore: use the underlying parse exception as the exception cause

Signed-off-by: TTtie <me@tttie.cz>

calcastor committed:
misc: use requireNonNullElse in BukkitUtils

Signed-off-by: BT (calcastor/mame) <43831917+calcastor@users.noreply.github.com>

TTtie committed:
refactor: don't use reflection for post-mortem interaction fix

Rewrites the post-mortem interaction block fix to not use reflection to set the appropriate fields.

As a side effect, this now fires a PlayerClientLoadedWorldEvent, slightly improving the player respawn simulation (albeit not fully).

Signed-off-by: TTtie <me@tttie.cz>
